File: backend/src/main/java/com/tutict/finalassignmentbackend/entity/PaymentRecord.java
package com.tutict.finalassignmentbackend.entity;

import com.baomidou.mybatisplus.annotation.*;
import lombok.Data;

import java.io.Serial;
import java.io.Serializable;
import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * 支付记录表实体类，对应数据库表 "payment_record"
 * 记录罚款的支付明细信息
 */
@Data
@TableName("payment_record")
public class PaymentRecord implements Serializable {
    /**
     * 序列化版本 UID
     */
    @Serial
    private static final long serialVersionUID = 1L;

    /**
     * 支付记录 ID，主键，自动增长
     */
    @TableId(value = "payment_id", type = IdType.AUTO)
    private Long paymentId;

    /**
     * 罚款记录 ID
     */
    @TableField("fine_id")
    private Long fineId;

    /**
     * 支付流水号
     */
    @TableField("payment_number")
    private String paymentNumber;

    /**
     * 支付金额 (元)
     */
    @TableField("payment_amount")
    private BigDecimal paymentAmount;

    /**
     * 支付方式 (Cash, BankCard, Alipay, WeChat, BankTransfer, Other)
     */
    @TableField("payment_method")
    private String paymentMethod;

    /**
     * 支付时间
     */
    @TableField("payment_time")
    private LocalDateTime paymentTime;

    /**
     * 支付渠道
     */
    @TableField("payment_channel")
    private String paymentChannel;

    /**
     * 缴款人姓名
     */
    @TableField("payer_name")
    private String payerName;

    /**
     * 缴款人身份证号
     */
    @TableField("payer_id_card")
    private String payerIdCard;

    /**
     * 缴款人联系电话
     */
    @TableField("payer_contact")
    private String payerContact;

    /**
     * 银行名称
     */
    @TableField("bank_name")
    private String bankName;

    /**
     * 银行账号
     */
    @TableField("bank_account")
    private String bankAccount;

    /**
     * 交易流水号
     */
    @TableField("transaction_id")
    private String transactionId;

    /**
     * 票据号码
     */
    @TableField("receipt_number")
    private String receiptNumber;

    /**
     * 票据文件 URL
     */
    @TableField("receipt_url")
    private String receiptUrl;

    /**
     * 支付状态 (Pending, Success, Failed, Refunded, Cancelled)
     */
    @TableField("payment_status")
    private String paymentStatus;

    /**
     * 退款金额 (元)
     */
    @TableField("refund_amount")
    private BigDecimal refundAmount;

    /**
     * 退款时间
     */
    @TableField("refund_time")
    private LocalDateTime refundTime;

    /**
     * 创建时间
     */
    @TableField(value = "created_at", fill = FieldFill.INSERT)
    private LocalDateTime createdAt;

    /**
     * 更新时间
     */
    @TableField(value = "updated_at", fill = FieldFill.INSERT_UPDATE)
    private LocalDateTime updatedAt;

    /**
     * 创建人
     */
    @TableField("created_by")
    private String createdBy;

    /**
     * 更新人
     */
    @TableField("updated_by")
    private String updatedBy;

    /**
     * 软删除时间戳
     */
    @TableField("deleted_at")
    @TableLogic(value = "null", delval = "now()")
    private LocalDateTime deletedAt;

    /**
     * 备注
     */
    @TableField("remarks")
    private String remarks;
}

File: flutter_app/lib/features/model/payment_record.dart
class PaymentRecordModel {
  final int? paymentId;
  final int? fineId;
  final String? paymentNumber;
  final double? paymentAmount;
  final String? paymentMethod;
  final DateTime? paymentTime;
  final String? paymentChannel;
  final String? payerName;
  final String? payerIdCard;
  final String? payerContact;
  final String? bankName;
  final String? bankAccount;
  final String? transactionId;
  final String? receiptNumber;
  final String? receiptUrl;
  final String? paymentStatus;
  final double? refundAmount;
  final DateTime? refundTime;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final String? createdBy;
  final String? updatedBy;
  final DateTime? deletedAt;
  final String? remarks;

  const PaymentRecordModel({
    this.paymentId,
    this.fineId,
    this.paymentNumber,
    this.paymentAmount,
    this.paymentMethod,
    this.paymentTime,
    this.paymentChannel,
    this.payerName,
    this.payerIdCard,
    this.payerContact,
    this.bankName,
    this.bankAccount,
    this.transactionId,
    this.receiptNumber,
    this.receiptUrl,
    this.paymentStatus,
    this.refundAmount,
    this.refundTime,
    this.createdAt,
    this.updatedAt,
    this.createdBy,
    this.updatedBy,
    this.deletedAt,
    this.remarks,
  });

  PaymentRecordModel copyWith({
    int? paymentId,
    int? fineId,
    String? paymentNumber,
    double? paymentAmount,
    String? paymentMethod,
    DateTime? paymentTime,
    String? paymentChannel,
    String? payerName,
    String? payerIdCard,
    String? payerContact,
    String? bankName,
    String? bankAccount,
    String? transactionId,
    String? receiptNumber,
    String? receiptUrl,
    String? paymentStatus,
    double? refundAmount,
    DateTime? refundTime,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? createdBy,
    String? updatedBy,
    DateTime? deletedAt,
    String? remarks,
  }) {
    return PaymentRecordModel(
      paymentId: paymentId ?? this.paymentId,
      fineId: fineId ?? this.fineId,
      paymentNumber: paymentNumber ?? this.paymentNumber,
      paymentAmount: paymentAmount ?? this.paymentAmount,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      paymentTime: paymentTime ?? this.paymentTime,
      paymentChannel: paymentChannel ?? this.paymentChannel,
      payerName: payerName ?? this.payerName,
      payerIdCard: payerIdCard ?? this.payerIdCard,
      payerContact: payerContact ?? this.payerContact,
      bankName: bankName ?? this.bankName,
      bankAccount: bankAccount ?? this.bankAccount,
      transactionId: transactionId ?? this.transactionId,
      receiptNumber: receiptNumber ?? this.receiptNumber,
      receiptUrl: receiptUrl ?? this.receiptUrl,
      paymentStatus: paymentStatus ?? this.paymentStatus,
      refundAmount: refundAmount ?? this.refundAmount,
      refundTime: refundTime ?? this.refundTime,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      createdBy: createdBy ?? this.createdBy,
      updatedBy: updatedBy ?? this.updatedBy,
      deletedAt: deletedAt ?? this.deletedAt,
      remarks: remarks ?? this.remarks,
    );
  }

  factory PaymentRecordModel.fromJson(Map<String, dynamic> json) {
    return PaymentRecordModel(
      paymentId: json['paymentId'],
      fineId: json['fineId'],
      paymentNumber: json['paymentNumber'],
      paymentAmount: _toDouble(json['paymentAmount']),
      paymentMethod: json['paymentMethod'],
      paymentTime: _parseDateTime(json['paymentTime']),
      paymentChannel: json['paymentChannel'],
      payerName: json['payerName'],
      payerIdCard: json['payerIdCard'],
      payerContact: json['payerContact'],
      bankName: json['bankName'],
      bankAccount: json['bankAccount'],
      transactionId: json['transactionId'],
      receiptNumber: json['receiptNumber'],
      receiptUrl: json['receiptUrl'],
      paymentStatus: json['paymentStatus'],
      refundAmount: _toDouble(json['refundAmount']),
      refundTime: _parseDateTime(json['refundTime']),
      createdAt: _parseDateTime(json['createdAt']),
      updatedAt: _parseDateTime(json['updatedAt']),
      createdBy: json['createdBy'],
      updatedBy: json['updatedBy'],
      deletedAt: _parseDateTime(json['deletedAt']),
      remarks: json['remarks'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'paymentId': paymentId,
      'fineId': fineId,
      'paymentNumber': paymentNumber,
      'paymentAmount': paymentAmount,
      'paymentMethod': paymentMethod,
      'paymentTime': paymentTime?.toIso8601String(),
      'paymentChannel': paymentChannel,
      'payerName': payerName,
      'payerIdCard': payerIdCard,
      'payerContact': payerContact,
      'bankName': bankName,
      'bankAccount': bankAccount,
      'transactionId': transactionId,
      'receiptNumber': receiptNumber,
      'receiptUrl': receiptUrl,
      'paymentStatus': paymentStatus,
      'refundAmount': refundAmount,
      'refundTime': refundTime?.toIso8601String(),
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
      'createdBy': createdBy,
      'updatedBy': updatedBy,
      'deletedAt': deletedAt?.toIso8601String(),
      'remarks': remarks,
    };
  }

  @override
  String toString() {
    return 'PaymentRecordModel(paymentId: $paymentId, fineId: $fineId, paymentAmount: $paymentAmount, paymentStatus: $paymentStatus)';
  }

  static List<PaymentRecordModel> listFromJson(List<dynamic> jsonList) {
    return jsonList
        .map((json) =>
            PaymentRecordModel.fromJson(json as Map<String, dynamic>))
        .toList();
  }

  static Map<String, PaymentRecordModel> mapFromJson(
      Map<String, dynamic> json) {
    final map = <String, PaymentRecordModel>{};
    if (json.isNotEmpty) {
      json.forEach((String key, dynamic value) {
        map[key] = PaymentRecordModel.fromJson(value as Map<String, dynamic>);
      });
    }
    return map;
  }

  static Map<String, List<PaymentRecordModel>> mapListFromJson(
      Map<String, dynamic> json) {
    final map = <String, List<PaymentRecordModel>>{};
    if (json.isNotEmpty) {
      json.forEach((String key, dynamic value) {
        map[key] =
            PaymentRecordModel.listFromJson(value as List<dynamic>);
      });
    }
    return map;
  }

  static DateTime? _parseDateTime(dynamic value) {
    if (value == null) return null;
    if (value is DateTime) return value;
    if (value is String && value.isNotEmpty) {
      return DateTime.tryParse(value);
    }
    return null;
  }

  static double? _toDouble(dynamic value) {
    if (value == null) return null;
    if (value is num) return value.toDouble();
    if (value is String && value.isNotEmpty) {
      return double.tryParse(value);
    }
    return null;
  }
}
